### PR TITLE
Update operations

### PR DESCRIPTION
**Pull request by kaifox**
Dear all,

finally a first proposal for the atomic update operations. 

There are still a few unclarities:
* What to do exactly in the wrapped property?
* Null treatment - does it look ok?
* Currently many methods in Property as default methods ... Could go to simple property,  - just would mean duplication in the WrappedProperty
* Tests are there for the basic behaviour (correct values retrieved and dispatched), but no new ones for concurrency aspects...
*

Sorry that I hit the organize imports key ... so there are more lines changed then necessary ...

### CODE REVIEW

**Review by kaifox — COMMENTED** [review]
small addition: fixing the equals method of mapbackeds

File: src/main/java/org/ossgang/commons/properties/WrapperProperty.java [hunk 1]
```diff
@@ -58,4 +60,10 @@ public T get() {
     public Subscription subscribe(Observer<? super T> listener, SubscriptionOption... options) {
         return observableValue.subscribe(listener, options);
     }
+
+    @Override
+    public Transition<T> accumulate(T x, BinaryOperator<T> accumulatorFunction) {
```
[the remark this answers is not in the document]
Comment (michi42) [reply src/main/java/org/ossgang/commons/properties/WrapperProperty.java:65]: I don't see it possible, at least not with reasonable guarantees.

What I could imagine though: We could leave the "Property" interface as it was, and introduce a new interface "AtomicProperty" that extends it and has all `accumulate` and all the atomic sugar methods.

I think this would be nice for a few reasons:
- basic "Property" stays a clean definition of "something you can set/get/subscribe" with no implications of whether these operations are atomic or not
- the default methods are nicely separated from the main Property interface
- WrapperProperty can implement `Property` (thus no accumulate) while SimpleProperty can implement `AtomicProperty`
- AtomicProperty is very explicit and close to AtomicReference :-)

What do you think?

File: src/test/java/org/ossgang/commons/observables/ExceptionHandlersTest.java [hunk 2]
```diff
@@ -61,7 +73,7 @@ public void nothingThrows_shouldNotCatchAnything() throws Exception {
     @Test(expected = TimeoutException.class)
     public void mapThrows_subscriberHandles_shouldNotDeflect() throws Exception {
         TestObserver<Integer> testObserver = new TestObserver<>();
-        Property<String> property = Properties.property("A");
+        Property<String> property = Properties.property();
```
Comment (michi42) [inline src/test/java/org/ossgang/commons/observables/ExceptionHandlersTest.java:76]: Out of curiosity - why these changes? I agree the initial value is not needed, but was there anything more behind it?

File: src/main/java/org/ossgang/commons/observables/DispatchingObservableValue.java [hunk 3]
```diff
@@ -51,18 +56,45 @@ public Subscription subscribe(Observer<? super T> listener, SubscriptionOption..
...
+            newValue.set(v);
+            return v;
+        });
+        Transition<T> transition = Transition.fromTo(oldValue, newValue.get());
+        dispatch(transition);
+        return transition;
+    }
+
+    private void dispatch(Transition<T> transition) {
+        if (Objects.equals(transition.oldValue(), transition.newValue())) {
```
Comment (michi42) [inline src/main/java/org/ossgang/commons/observables/DispatchingObservableValue.java:94]: Probably this we could put on `Transition` (in a method called `hasChange()` or so)? 

Just cosmetics of course.

File: src/main/java/org/ossgang/commons/observables/DispatchingObservableValue.java [hunk 4]
```diff
@@ -51,18 +56,45 @@ public Subscription subscribe(Observer<? super T> listener, SubscriptionOption..
...
-            Optional.ofNullable(lastValue.get()).ifPresent(uncheckedConsumer(value -> dispatch(listener::onValue, value).get()));
+            Optional.ofNullable(lastValue.get())
+                    .ifPresent(uncheckedConsumer(value -> dispatch(listener::onValue, value).get()));
         }
         return subscription;
     }
 
     @Override
     protected void dispatchValue(T newValue) {
         requireNonNull(newValue, "null value not allowed");
```
Comment (michi42) [inline src/main/java/org/ossgang/commons/observables/DispatchingObservableValue.java:67]: This check we can remove - accumulate does it.

File: src/main/java/org/ossgang/commons/observables/Transition.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+ * properties.
+ */
+public class Transition<T> {
+
+    private final T oldValue;
+    private final T newValue;
+
+    private Transition(T oldValue, T newValue) {
+        this.oldValue = oldValue; /* The old value is allowed to be null */
+        this.newValue = requireNonNull(newValue, "new value must not be null");
```
Comment (michi42) [inline src/main/java/org/ossgang/commons/observables/Transition.java:19]: I would not enforce null or non-null here - I think ideally the knowledge that the previous value can be null (in the very special case of an uninitialized observable) but the updated value can should be in `DispatchingObservableValue` and nowhere else.

**Comment by michi42** [comment]
Hi @kaifox, apologies that this took me ages to have a look at - we are quite busy at CERN with all the start up...

Most comments are cosmetics, but the `Property` vs `AtomicProperty` we should discuss.

File: src/test/java/org/ossgang/commons/observables/ExceptionHandlersTest.java (hunk 2 above)
Comment (kaifox) [reply src/test/java/org/ossgang/commons/observables/ExceptionHandlersTest.java:76]: Honestly ...not even sure if I did the change ... I fear that I did some mess during the rebase ... so could it be that this change comes from some of andreas changes ... Sorry.

File: src/main/java/org/ossgang/commons/observables/DispatchingObservableValue.java (hunk 3 above)
Comment (kaifox) [reply src/main/java/org/ossgang/commons/observables/DispatchingObservableValue.java:94]: indeed

**Comment by kaifox** [comment]
Hi, no worries. Yes I know ... there is a lot going on. I will think about the atomic property thing (have to go back, as it is quite some time ago I did this ...)

File: src/main/java/org/ossgang/commons/observables/Transition.java (hunk 5 above)
Comment (kaifox) [reply src/main/java/org/ossgang/commons/observables/Transition.java:19]: Ok. Sounds reasonable.

File: src/main/java/org/ossgang/commons/observables/DispatchingObservableValue.java (hunk 4 above)
Comment (kaifox) [reply src/main/java/org/ossgang/commons/observables/DispatchingObservableValue.java]: Changed in latest commit

File: src/main/java/org/ossgang/commons/observables/Transition.java (hunk 5 above)
Comment (kaifox) [reply src/main/java/org/ossgang/commons/observables/Transition.java]: Done
